java/src/main/java/ai/rapids/cudf/NvtxRange.java

```diff
@@ -33,32 +37,89 @@
  *
  * Instances should be associated with a single thread to avoid pushing an NVTX range in
  * one thread and then trying to pop the range in a different thread.
+ *
+ * Push/pop ranges show a stacking behavior in tools such as Nsight, where newly pushed 
+ * ranges are correlated and enclosed by the prior pushed range (in the example above,
+ * "b" is enclosed by "a").
+ *
+ * Start/end:
+ *
+ * The constructor instantiates a new NVTX range and keeps a handle that comes back from the
+ * NVTX api (nvtxRangeId) that used to later close such a range. This type of range does 
+ * not have the same order-of-operation requirements that the push/pop ranges have: 
+ * the `NvtxRange` instance can be passed to other scopes, and even to other threads 
+ * for the eventual call to close.
+ *
+ * It can be used in the same try-with-resources way as push/pop, or interleaved with other
+ * ranges, like so:
+ *
+ * <pre>
+ *   NvtxRange a = new NvtxRange("a", NvtxColor.RED, NvtxRange.Type.STARTEND);
+ *   NvtxRange b = new NvtxRange("b", NvtxColor.BLUE, NvtxRange.Type.STARTEND);
+ *   a.close();
+ *   b.close();
+ * </pre>
+ *
+ * Start/end ranges are different in that they don't have the same correlation that the
+ * push/pop ranges have. 
  */
 public class NvtxRange implements AutoCloseable {
+  public enum Type {
+    PUSH,
+    STARTEND
```

Comment:
One difference I see is the push/pop range is visible right away in the nsys-ui:

![twisty_1](https://user-images.githubusercontent.com/1901059/139685849-c0327c83-d26c-479e-8d40-fbe469d1ae4d.png)

You have to click a twisty to see the start/end ranges:

![twisty_2](https://user-images.githubusercontent.com/1901059/139685856-dbabbd8c-cec9-4506-8579-79b002367931.png)

This seems small, but honestly, from the user's perspective it means an extra click for something that would otherwise be readily visible (per thread). I propose we keep both, using the push/pop in the stack-driven approach, and keep the start/end ones for more complex ranges that span threads or where ranges get passed around. Otherwise, the correlation behavior works (for an enclosed start/end range), so yeah it's very similar.

As an experiment, I turned every range into a start/end range and all of them were hidden by default.

Reply:
@jlowe let me know if given the above you are OK with keeping both impls. If so, I can work on breaking it up into two classes if that's the route we want to take.

Reply:
Given your findings, agree we need the two implementations, so that leaves the question of how to advertise the two different implementations.

I have a preference for two different classes given the existing prototype shows there's almost no shared code between the two.  I don't see a major advantage of forcing the two separate implementations under the existing `NvtxRange` class unless someone wanted to wield both types of ranges under a common interface name and needed more specific type information than the existing `AutoCloseable` interface.  We could always add a new interface they both share if that ends up being a needed use-case.

How about calling the separate, new range class something like `NvtxUnscopedRange`?  The existing one should be used when it is tied to a code scope, guaranteeing it will close before returning to the parent scope, while the new one supports closing after leaving the scope in which it was created.

Reply:
I split this up into a different class @jlowe. I reverted the double-close check in `NvtxRange` fyi, just to not introduce a change this late in 21.12 that could break existing code. We can make those changes after 22.02 is available.

Reply:
fwiw, the naming of the corresponding C++ objects are `scoped_range` and `unique_range`.

Reply:
I'm fine with calling the new class `NvtxUniqueRange` to be more inline with the C++ naming.